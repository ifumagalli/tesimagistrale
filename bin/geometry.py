""" Classes and functions for geometry managing.
	Everything is done for a 2D domain.
"""

from dolfin import *
from numpy import array

# ===== Boundaries =====

class MovingBoundary(SubDomain):
	def inside(self,x,on_boundary):
		return on_boundary and near(x[1],0)

class StillBoundary(SubDomain):
    def __init__(self,moving_boundary):
        SubDomain.__init__(self)
        self._moving_boundary=moving_boundary
	def inside(self,x,on_boundary):
		return on_boundary and not self._moving_boundary.inside(x,on_boundary)


# ===== MapTq: Omega_0->Omega_q =====

class MapTq(Expression):
	""" Class implementing the map Tq: Omega_0->Omega_q
		which maps the reference domain into the actual one.
		It depends on the control function q, which identifies Omega_q.
		! Cartesian 2D domains are considered
	"""
	def __init__(self,*args,**kwargs):
		# NB Arguments have to be passed using **kwargs (e.g.   qh=Function(Qh)
		#														Tq = MapTq(q=qh)) 
		#	 and Expression.__init__(...) has not to be called directly: see the link below
		# http://fenicsproject.org/documentation/dolfin/1.2.0/python/programmers-reference/functions/
		#	expression/Expression.html#dolfin.functions.expression.Expression
		self.q = kwargs.get('q')
			# control: a scalar Function on an Interval

	def eval(self,values,x):
		""" Given a point x=[x[0] x[1]] in the REFERENCE domain Omega_0
			it returns in 'values' the point Tq(x) in the ACTUAL domain Omega_q.
			As far, Omega_q is similar to Omega_0, with the bottom border substituted by q's graph.
		"""
		values[0] = x[0]
		tmp = self.q(x[0])
		values[1] = x[1]+(1-x[1])*tmp
		
	def value_shape(self):
		return (2,)
			# we consider 2D domains
		
	def inverse(self,value,x):
		""" Given a point x=[x[0] x[1]] in the ACTUAL domain Omega_q
			it returns in 'values' the point Tq^{-1}(x) in the REFERENCE domain Omega_0
		"""
		value = (x[0], (x[1]-self.q(x[0]))/(1-self.q(x[0])))
