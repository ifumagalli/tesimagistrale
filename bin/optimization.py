""" Classes implementing optimization methods.
    NB ??? Per ora GradientMethod eredita da OptimizationMethod, ma essendo una policy forse sara' meglio cambiare il legame tra le due classi
"""

from math import *
from functions import *
import copy

#=====================

class OptimizationMethod(object):
	""" Abstract class defining an iterative method for optimize a "function" f(x)
		starting from x0, with a tolerance and a maximum number of iterations
	"""
	def __init__(self,tol,maxit):
		# self._f = f
		# self._x0 = x0
		self._tol = tol
		self._maxit = maxit
		# serviranno altre tolleranze in altri oggetti
		# self._iter = 0
		# self._xk = x0
		# self._fxk = self._f(x0)
		# self._xold = None
		# self._fxold = None

	def apply(self,f,x0):
		pass
	
	# stop criteria
	def stop_normx(self):
		iter_check = (self._iter > self._maxit)
		if iter_check:
			print "  ***MAXIMUM NUMBER OF ITERATIONS EXCEEDED***"
			print "  ***in OptimizationMethod.stop_normx()***"
			return iter_check
		# return iter_check or (sum(abs(self._xk-self._xold)) < self._tol)
		return iter_check or (errornorm(self._xk,self._xold,'l2') < self._tol)
			# ??? come fare per tenere generalizzazione al caso di x vettoriale? Perche' gli scalari non hanno un metodo length
		# returns True if stop is needed
	
	_stop_criterion = stop_normx

	# TODO
	# ...set,get,info...

#=====================
	
class GradientMethod(OptimizationMethod):
	""" The "function" f(x) to be optimized need to be a class with a method derivative(x) that returns the derivative of f applied in x.
		With 'derivative' we mean the gradient or, in general, the Riesz element of the (Frechet-)differential
	"""
	def __init__(self,tol,maxit):
		OptimizationMethod.__init__(self,tol,maxit)
		# self._alphak = _alpha0
		# self._gradf = gradf		# TODO il gradiente sara' un metodo dell'oggetto f

	def apply(self,f,x0,alpha0=1.0):
		self._xk = x0
		self._fxk = f(x0)
		#f.eval(self._fxk,x0)
		self._alphak = alpha0
		self._xold = x0
		#self._xold = 0*x0
		#self._xold = copy.deepcopy(self._xk)
		#self._xold = x0.__init__(x0)
		#self._xold = copy.deepcopy(x0)
		#self._xold.vector()[:] *= DOLFIN_EPS
			# ??? metto 0*stesso_oggetto affinche' funzioni il criterio di arresto, ma preferirei un bel None
			# !!! e non posso neanche mettere 0* perche' altrimenti mi diventa _xold=Zero e l'oggetto Zero non ha un metodo vector()
		self._fxold = 0*self._fxk	# idem
		self._iter = 0
		while self._iter==0 or (not self._stop_criterion()):
		# while self._iter==0 or conditional(Not(self._stop_criterion()),True,False):
			# ??? non c'e' un modo per far si' che _stop_criterion restituisca un bool di Python?
			# perche' usando la riga sopra commentata da' l'errore seguente: 
			# " UFL conditions cannot be evaluated as bool in a Python context "
			self._iter += 1
			print "iter =",self._iter
			print "xk =",self._xk,"xold = ",self._xold
			print "f(xk) =",self._fxk,"f(xold) =",self._fxold
			print "f'(xk) =",f.gradient(self._xk),"f'(xold) =",f.gradient(self._xold)
			# print abs(self._xk-self._xold),self._stop_criterion()
			# print "-------"
			# self._xold = copy.deepcopy(self._xk)	# ??? DEEPCOPY
			self._xold = self._xk					# shallow copy, but it's fine
			self._alphak = self._update_alphak()
			self._xk = self._xk - (self._alphak*f.gradient(self._xk))	# DO NOT use -= because so far _xold and _xk share the same object
			self._fxold = self._fxk					# shallow copy, but it's fine
			self._fxk = f(self._xk)
			# print self._xk,self._xold,abs(self._xk-self._xold),self._stop_criterion()
	# DEBUG	
			# plot(self._xk,mesh=UnitIntervalMesh(20),title="qk(x)")
			print [self._xk.vector()[i] for i in range(0,len(self._xk.vector()))]
			print [self._xold.vector()[i] for i in range(0,len(self._xold.vector()))]
			print numpy.abs(self._xk.vector()-self._xold.vector())
			raw_input("Press ENTER to continue")
			print "-------"
		return self._xk
	
	# TODO updating techniques
	# def linear_search(self,...):
	#	...
	#	return alphak
	# per il momento non aggiorniamo, con il seguente
	def _update_alphak(self):
		return self._alphak
	
	# ...set,get,info...


#############################################################################
# class OptimizationAlgorithm:
	# def __init__(self,optTol,stateSolver,optMethod,linearSystemTol=1.e-3*optTol):
		# self._optTol = optTol
			# # serviranno anche altre tolleranze come linearSystemTol
		# self._stateSolver = stateSolver
		# self._optMethod = optMethod
			# # Has to be an OptimizationMethod
		
	# def setMethod(self,optMethod):
		# self._optMethod = optMethod
	# def getMethod(self):
		# return self._optMethod
	# def infoMethod(self):
		# return self._optMethod.info
	# # ...
	# # analoghi metodi per gli altri attributi della classe
