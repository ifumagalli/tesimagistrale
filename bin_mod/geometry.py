""" Classes and functions for geometry managing
	Everything is done for a 2D domain.
"""

from dolfin import *
from numpy import array

# ===== Boundaries =====

class MovingBoundary(SubDomain):
	def inside(self,x,on_boundary):
		return on_boundary and x[1] < DOLFIN_EPS

class StillBoundary(SubDomain):
    def __init__(self,moving_boundary):
        SubDomain.__init__(self)
        self._moving_boundary=moving_boundary
	def inside(self,x,on_boundary):
		return on_boundary and not self._moving_boundary.inside(x,on_boundary)


# ===== MapTq: Omega->Omega_q =====

class MapTq(Expression):	# inheriting from class dolfin.functions.expression.Expression
	def __init__(self,*args,**kwargs):
		# NB Arguments have to be passed using **kwargs (e.g.   qh=Function(Qh)
		#														Tq = MapTq(q=qh)) 
		#	 and Expression.__init__(...) has not to be called directly: see the link below
		# http://fenicsproject.org/documentation/dolfin/1.2.0/python/programmers-reference/functions/
		#	expression/Expression.html#dolfin.functions.expression.Expression
		self.q = None	# control: a scalar Function on an Interval
		# TODO: completare

	# def set_q(self,q):
		# self.q = q

	def eval(self,values,x):
		""" Given a point x=[x[0] x[1]] in the REFERENCE domain Omega_0
			it returns in 'values' the point Tq(x) in the ACTUAL domain Omega_q
		"""
		values[0] = x[0]
		print (q)
		tmp = self.q(x[0])
		values[1] = x[1]+(1-x[1])*tmp
		
	def value_shape(self):
		return (2,)
		
	def inverse(self,value,x):
		""" Given a point x=[x[0] x[1]] in the ACTUAL domain Omega_q
			it returns in 'values' the point Tq^{-1}(x) in the REFERENCE domain Omega_0
		"""
		value = (x[0], (x[1]-self.q(x[0]))/(1-self.q(x[0])))	

	def move_back(self,u):
		""" Given a function u defined on the ACTUAL domain Omega_q
			it returns u composed with Tq^{-1}
		"""
		TODO
		#MA FORSE NON E' DA FARE: E' SEMPLICEMENTE UNA COMPOSIZIONE (v. Function in Fenics)
	
	# class Jac(Expression):
		
		# def eval(self,values,x):
			# values = ((1,0)
		
		# def value_shape(self):
			# return (2,2)

#=======
# other objects deriving from MapTq
# ??? vorrei fare l'overloading dell'operatore nabla_grad, ma non so come e' fatto, e poi in Python non esiste overloading

# class DTq(LinearOperator):
	# def __init__(self,q)
		# vector = Vector(2)
		# LinearOperator.__init__(self, vector, vector)

#	# matrix-vector product
	# def mult(self, src, dst):
		# self.B.transpmult(src, self.tmp1)
		# self.Ainv.mult(self.tmp1, self.tmp2)
		# self.B.mult(self.tmp2, dst)
	# MEGLIO FARLO DERIVARE DA Expression
#class DTq(Expression):
# class DTq(Expr): # from ufl
	# def __init__(self,q):
		# Expression.__init__(self,2,2)
		# self.q = q
	
	# def eval(self,values,x):
		# values = np.array([[1 , 0] , [(1-x[1])*grad(q)(x[0]) , 1-q(x[0])]])
	
	# def inverse(self,values,x):
		# values = np.array([[1 , 0] , [-(1-x[1])/(1-q(x[0]))*grad(q)(x[0]) , 1.0/(1-q(x[0]))]])
