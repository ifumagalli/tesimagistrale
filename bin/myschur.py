#!/usr/bin/env python

from dolfin import *
from geometry import *

#parameters['linear_algebra_backend'] = 'PETSc'

# class IterativeInverse(LinearOperator):
  # # constructor
  # def __init__(self, vector, matrix, solver, preconditioner=None):
    # LinearOperator.__init__(self, vector, vector)
    # self.matrix = matrix
    # self.solver = solver
    # self.preconditioner = preconditioner

    # if preconditioner is None:
      # self.solver.set_operator(self.matrix)
    # else:
      # self.solver.set_operators(self.matrix, self.preconditioner)

  # def size(self, dim):
    # return self.matrix.size(dim)

  # # matrix-vector product
  # def mult(self, src, dst):
    # self.solver.solve(dst, src)


# class SchurComplement(LinearOperator):
  # # constructor
  # def __init__(self, vector, Ainv, B):
    # LinearOperator.__init__(self, vector, vector)
    # self.Ainv = Ainv
    # self.B = B

    # self.tmp1 = Vector(Ainv.size(1));
    # self.tmp2 = Vector(Ainv.size(1));

  # # matrix-vector product
  # def mult(self, src, dst):
    # self.B.transpmult(src, self.tmp1)
    # self.Ainv.mult(self.tmp1, self.tmp2)
    # self.B.mult(self.tmp2, dst)


class StokesSolver(object):
  # constructor
  def __init__(self, Vh, Qh):
    self.Vh = Vh
    self.Qh = Qh
    self.Wh = Vh * Qh

  # assemble matrix
  def assemble(self, mu, f, bc, control):
    # # Dirichlet bc
    # # u0_boundary = lambda x, on_boundary : on_boundary
    # # bc = DirichletBC(self.Vh, u0, u0_boundary)
    # Gamma_moving = MovingBoundary()
    # Gamma_still = StillBoundary(Gamma_moving)
    # # Dirichlet bc
    # u_in = Expression(('sqrt(x[1])','0'))		# inflow velocity
    # noslip = DirichletBC(self.Vh,Expression(('0','0')),lambda x, on_boundary : Gamma_moving.inside(x,on_boundary))
    # inflow = DirichletBC(self.Vh,u_in,lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and x[0] < DOLFIN_EPS)
    # # noslip2 = DirichletBC(V,Expression(('0','0')),lambda x, on_boundary : Gamma_still.inside(x,on_boundary) and x[1] > 1-DOLFIN_EPS)
    # symmetry = DirichletBC(self.Vh.sub(1),Expression('0'),lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and x[1] > 1-DOLFIN_EPS)
	# # ??? DA VERIFICARE
    # bc = [noslip,inflow,symmetry]


    u,p = TrialFunctions(self.Wh)
    v,q = TestFunctions(self.Wh)

    # coefficients due to control
    Tq = MapTq(element=self.Vh.ufl_element())
    Tq.q = control
    DTq = grad(Tq)
    gammaq = det(DTq)
    DTq_invT = transpose(inv(DTq))
    Aq = gammaq*DTq*DTq_invT

    # assemble A matrix
    # eps = lambda u: sym(grad(u))
    # a = 2 * mu * inner(eps(u), eps(v)) * dx
    a = mu*inner(grad(u)*Aq,grad(v)) * dx \
      - inner(grad(u),DTq_invT)*q*gammaq * dx \
      - inner(grad(v),DTq_invT)*p*gammaq * dx
    L = dot(f, v)*gammaq * dx
	# TODO inserire dato di Neumann
    self.A, self.b = assemble_system(a, L, bc)

    # assemble A preconditioner
    # !!! tengo lo stesso...
    a = mu * inner(grad(u),grad(v)) * dx \
	  + p * q / mu * dx
    self.Ap = assemble_system(a, L, bc)[0]

  def solve(self):
	x = Function(self.Wh)
	
	#solver = KrylovSolver('gmres', 'ilu')
	#solver.set_operators(self.A, self.Ap)
	solver = LUSolver()
	solver.set_operator(self.A)
	solver.solve(x.vector(), self.b)
	
	return x.split()
	
  def mesh(self):
	return self.Vh.mesh()

# =====================

# N = 10
# mesh = UnitSquareMesh(N,N,'crossed')

# Vh = VectorFunctionSpace(mesh, 'CG', 2)
# Qh = FunctionSpace(mesh, 'CG', 1)

# stokes = StokesSolver(Vh, Qh)

# # viscosity and force
# mu = Constant(1)
# f = Expression(('0.5 * sin(2*x[0]) - 2 * cos(x[0]) * sin(x[1])',
#                '0.5 * sin(2*x[1]) + 2 * sin(x[0]) * cos(x[1])'))

# # solution
# u0 = Expression(('- cos(x[0]) * sin(x[1])',
#                'sin(x[0]) * cos(x[1])'))
# p0 = Expression('0.25 * (cos(2*x[0]) + cos(2*x[1]))')

# stokes.assemble(mu, f, u0)
# u,p = stokes.solve()

# u_x,u_y = u.split()
# plot(u,title=("Velocity vector u"))
# plot(u_x,title=("Velocity x component"))
# plot(u_y,title=("Velocity y component"))
# plot(p,title=("Pressure"))
# interactive()

