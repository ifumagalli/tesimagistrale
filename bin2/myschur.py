#!/usr/bin/env python

from dolfin import *
from geometry import *

parameters['linear_algebra_backend'] = 'PETSc'

class IterativeInverse(LinearOperator):
  # constructor
  def __init__(self, vector, matrix, solver, preconditioner=None):
    LinearOperator.__init__(self, vector, vector)
    self.matrix = matrix
    self.solver = solver
    self.preconditioner = preconditioner

    if preconditioner is None:
      self.solver.set_operator(self.matrix)
    else:
      self.solver.set_operators(self.matrix, self.preconditioner)

  def size(self, dim):
    return self.matrix.size(dim)

  # matrix-vector product
  def mult(self, src, dst):
    self.solver.solve(dst, src)


class SchurComplement(LinearOperator):
  # constructor
  def __init__(self, vector, Ainv, B):
    LinearOperator.__init__(self, vector, vector)
    self.Ainv = Ainv
    self.B = B

    self.tmp1 = Vector(Ainv.size(1));
    self.tmp2 = Vector(Ainv.size(1));

  # matrix-vector product
  def mult(self, src, dst):
    self.B.transpmult(src, self.tmp1)
    self.Ainv.mult(self.tmp1, self.tmp2)
    self.B.mult(self.tmp2, dst)


class StokesSolver(object):
  # constructor
  def __init__(self, Vh, Qh):
    self.Vh = Vh
    self.Qh = Qh

  # assemble matrix
  def assemble(self, mu, f, bc, ctrl):
    # # Dirichlet bc
    # # u0_boundary = lambda x, on_boundary : on_boundary
    # # bc = DirichletBC(self.Vh, u0, u0_boundary)
    # Gamma_moving = MovingBoundary()
    # Gamma_still = StillBoundary(Gamma_moving)
    # # Dirichlet bc
    # u_in = Expression(('sqrt(x[1])','0'))		# inflow velocity
    # noslip = DirichletBC(self.Vh,Expression(('0','0')),lambda x, on_boundary : Gamma_moving.inside(x,on_boundary))
    # inflow = DirichletBC(self.Vh,u_in,lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and x[0] < DOLFIN_EPS)
    # # noslip2 = DirichletBC(V,Expression(('0','0')),lambda x, on_boundary : Gamma_still.inside(x,on_boundary) and x[1] > 1-DOLFIN_EPS)
    # symmetry = DirichletBC(self.Vh.sub(1),Expression('0'),lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and x[1] > 1-DOLFIN_EPS)
	# # ??? DA VERIFICARE
    # bc = [noslip,inflow,symmetry]


    u = TrialFunction(self.Vh)
    p = TrialFunction(self.Qh)

    v = TestFunction(self.Vh)
    q = TestFunction(self.Qh)

    # coefficients due to control
    #@#Tq = MapTq(element=self.Vh.ufl_element())
    #@#Tq.q = ctrl
    #@#DTq = grad(Tq)
    #@#gammaq = det(DTq)
    #@#DTq_invT = transpose(inv(DTq))
    #@#Aq = gammaq*DTq*DTq_invT

    # assemble A matrix
    # eps = lambda u: sym(grad(u))
    # a = 2 * mu * inner(eps(u), eps(v)) * dx
    a = mu*inner(grad(u),grad(v)) * dx
    #@#a = mu*inner(grad(u)*Aq,grad(v)) * dx
    L = dot(f, v) * dx
    #@#L = dot(f, v)*gammaq * dx
	# TODO inserire dato di Neumann
    self.A, self.b = assemble_system(a, L, bc)

    # assemble A preconditioner
    # !!! tengo lo stesso...
    a = mu * inner(grad(u),grad(v)) * dx
    self.Ap = assemble_system(a, L, bc)[0]

    # assemble B matrix
    a = - div(u) * q * dx
    #@#a = - inner(grad(u),DTq_invT)*q*gammaq * dx
    self.B = assemble(a)

    # assemble Schur preconditioner
    a = p * q / mu * dx
    self.M = assemble(a)

  def solve(self):
    u = Function(self.Vh)
    p = Function(self.Qh)

    #solver1 = KrylovSolver('gmres', 'none')
    #Ainv = IterativeInverse(u.vector(), self.A, solver1, self.Ap)
    solver1 = LUSolver()
    Ainv = IterativeInverse(u.vector(), self.A, solver1)

    S = SchurComplement(p.vector(), Ainv, self.B)
    solver2 = KrylovSolver('gmres', 'ilu')
    solver2.set_operators(S, self.M)

    tmp1 = Vector(u.vector().size())
    tmp2 = Vector(p.vector().size())

    Ainv.mult(self.b, tmp1)
    self.B.mult(tmp1, tmp2)

    solver2.solve(p.vector(), tmp2)

    self.B.transpmult(p.vector(), tmp1)
    tmp1 -= self.b
    tmp1 *= -1

    Ainv.mult(tmp1, u.vector())

    return u,p

# =====================

# N = 10
# mesh = UnitSquareMesh(N,N,'crossed')

# Vh = VectorFunctionSpace(mesh, 'CG', 2)
# Qh = FunctionSpace(mesh, 'CG', 1)

# stokes = StokesSolver(Vh, Qh)

# # viscosity and force
# mu = Constant(1)
# f = Expression(('0.5 * sin(2*x[0]) - 2 * cos(x[0]) * sin(x[1])',
#                '0.5 * sin(2*x[1]) + 2 * sin(x[0]) * cos(x[1])'))

# # solution
# u0 = Expression(('- cos(x[0]) * sin(x[1])',
#                'sin(x[0]) * cos(x[1])'))
# p0 = Expression('0.25 * (cos(2*x[0]) + cos(2*x[1]))')

# stokes.assemble(mu, f, u0)
# u,p = stokes.solve()

# u_x,u_y = u.split()
# plot(u,title=("Velocity vector u"))
# plot(u_x,title=("Velocity x component"))
# plot(u_y,title=("Velocity y component"))
# plot(p,title=("Pressure"))
# interactive()

