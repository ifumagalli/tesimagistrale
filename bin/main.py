from dolfin import *
from optimization import *
from geometry import *
from numpy import array
from myschur import StokesSolver
	# ??? dove viene usata la PETSc

# my_debug = True
my_debug = False

# info(parameters,True)
parameters["allow_extrapolation"] = True	# useful for evaluating on boundaries (see footnote on p.116 of the manual)

# mesh and spaces definitions
Nq = 20
I = UnitInterval(Nq) # mesh for controls
# I = UnitIntervalMesh(Nq) # mesh for controls !!! better: previous is deprecated
Nup = Nq
    # ??? ci vuole un legame tra Nq e Nup? Non direi...
    # ??? idem per i gradi polinomiali
Omega = UnitSquare(Nup,Nup) # mesh for velocity and pressure on reference domain
# Omega = UnitSquareMesh(Nup,Nup) # mesh for velocity and pressure on reference domain !!! better: previous is deprecated
degq = 2	# polynomial degree for controls FE space Q
	# every q in Q is supposed to be in H2, then we guarantee not to have necessarily q''=0 setting degq>=2
degu = 2	# polynomial degree for velocity FE space V
degp = 1	# polynomial degree for pressure FE space P
Q = FunctionSpace(I,'CG',degq)
Q_Omega = FunctionSpace(Omega,'CG',degq)
V = VectorFunctionSpace(Omega,'CG',degu)
P = FunctionSpace(Omega,'CG',degp)
Vexpr = V	# used by Expressions: maybe a high order is better

# parameters for the optimization method
# tol = 1.0e-3
tol = DOLFIN_EPS
maxit = 2
alpha0 = 0.5 # for descent methods
# penalty coefficients
lambda1 = 1		# for volume conservation
lambda2 = 1		# for q'' penalization

# data for the state problem
nu = Constant(1)	# viscosity
eta = Constant(0)		# 0-order parameter for generalized Stokes problem
f = Expression(('0.0','0.0'))		# external force
# for boundary conditions
u_in = Expression(('sqrt(x[1])','0'))		# inflow velocity
stress_out = Expression(('0','0'))	        # outflow stress

# functions and initial settings
qt = TrialFunction(Q)
ut = TrialFunction(V)
pt = TrialFunction(P)
zt = TrialFunction(V)
st = TrialFunction(P)
q = Function(Q)
u = Function(V)
z = Function(V)
s = Function(P)
#q0 = Expression('0.2-0.2*4*(x[0]-0.5)*(x[0]-0.5)')
q0 = Expression('sin(2*pi*x[0])')
u0 = Expression(('0.0','0.0'))
# q.vector()[:] = array([q0(x) for x in I.coordinates()])
	# !!! the line above works only for polynomial degree = 1
q = project(q0,Q)

# boundary managing
# boundary_parts = MeshFunction("uint", Omega, Omega.topology().dim()-1)
# boundary_parts = MeshFunction("size_t", Omega, Omega.topology().dim()-1) # uint as MeshFunction type has been deprecated since DOLFIN 1.1.0
	# ??? sarebbe meglio usare questa MeshFunction, invece dei SubDomain sotto definiti?
Gamma_moving = MovingBoundary()
# Gamma_still = StillBoundary()
# Dirichlet bc
noslip = DirichletBC(V,Expression(('0','0')),lambda x, on_boundary : Gamma_moving.inside(x,on_boundary))
inflow = DirichletBC(V,u_in,lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and near(x[0],0))#x[0] < DOLFIN_EPS)
# noslip2 = DirichletBC(V,Expression(('0','0')),lambda x, on_boundary : Gamma_still.inside(x,on_boundary) and x[1] > 1-DOLFIN_EPS)
symmetry = DirichletBC(V.sub(1),Expression('0'),lambda x, on_boundary : (not Gamma_moving.inside(x,on_boundary)) and near(x[1],1))#x[1] > 1-DOLFIN_EPS)
	# ??? DA VERIFICARE
bc = [noslip,inflow,symmetry]


# coefficients due to transformation on reference domain
# DEVONO ESSERE ESPRESSIONI CHE DIPENDONO DAL CONTROLLO
Tq = MapTq(element=Vexpr.ufl_element())
	# ??? passandogli questo element ho il grado di interpolazione dato da Vexpr: non e' che facendo il gradiente poi perdo un grado?
Tq.q = q0
DTq = grad(Tq)
gammaq = det(DTq)
DTq_invT = transpose(inv(DTq))
Aq = gammaq*DTq*DTq_invT

if my_debug == True:
	#u.vector()[:] = as_vector([Tq(Omega.coordinates()[i]) for i in range(0,len(Omega.coordinates()))])
	plot(q,mesh=I,title="q(x)")
	plot(Tq,mesh=Omega,title="Tq(x,y)")
	plot(Tq-Expression(('x[0]','x[1]')),mesh=Omega,title="Tq(x,y) - I")
	plot(gammaq,mesh=Omega,title="gamma_q(x,y)")
	interactive()

# mesh utilities
n = FacetNormal(Omega)

# cost functional J and its gradient
# FOR DERIVATIVES WE COULD USE UFL
cost_functional_form = 0.5*inner(grad(u)*Aq,grad(u))*dx
# penalty_terms = 0.5*lambda1*((q*dx) - Vbar)**2 \
#				+ 0.5*lambda2*Dx(Dx(q,0),0)*Dx(Dx(q,0),0)*dx
	# ??? non so se il primo pezzo funziona, scritto cosi'... per ora proviamo il seguente
    # quando decommenti sopra, decommenta anche penalty_gradient!!
penalty_terms = 0.5*lambda2*Dx(Dx(q,0),0)*Dx(Dx(q,0),0)*dx
J_form = cost_functional_form #+ penalty_terms
# J_form = (u - u0)*(u - u0)*dx
### J_form = dot(u-u0,u-u0)*dx
    # TODO per ora teniamo questo funzionale elementare, per vedere se le cose vanno...
cost_functional_gradient = - 0.5*inner(grad(u)*DTq,grad(u)*DTq)*ds(0)\
							 - dot(grad(u)*DTq*DTq_invT*n, (grad(u)-nu*grad(z))*DTq*DTq_invT*n + s*n)*ds(0)
# penalty_gradient = lambda1*((q*dx)-Vbar) + lambda2*Dx(Dx(Dx(Dx(q,0),0),0),0)*dx
penalty_gradient = lambda2*Dx(Dx(Dx(Dx(qt,0),0),0),0)*dx
	# ??? va bene con la derivata 4'?
gradJ_form = cost_functional_gradient #+ penalty_gradient
### gradJ_form = -inner(grad(z), w0*grad(u))*dx

# initializations
optmethod = GradientMethod(tol,maxit)
solver = StokesSolver(V,P)
#solver.assemble(nu,f,bc,q)
	# TODO per ora lasciamo questa Expression: poi ci andranno le bc

# functional object definition
J = CostFunctional(nu,f,bc,J_form,solver,grad_form=gradJ_form,q0=q0)
# apply optimization method
# optmethod.set_gradf(gradJ)	# TODO per ora lo mettiamo qui, ma dev'esserci un metodo gradiente nella classe di J
				# DONE messo in CostFunctional
q_opt = optmethod.apply(J,q)
	# ! Be sure to have set q to its initial value
u_opt,p_opt = J.last_state()

# post-processing
print "\nPost-processing"
# plot(q_opt,mesh=I,title="q(x)")
# plot(Tq,mesh=Omega,title="Tq(x,y)")
# plot(Tq-Expression(('x[0]','x[1]')),mesh=Omega,title="Tq(x,y) - I")
# plot(gammaq,mesh=Omega,title="gamma_q(x,y)")
# interactive()
raw_input("Press ENTER to continue")

print ""
print "Bien: On a fini Normalement!"
