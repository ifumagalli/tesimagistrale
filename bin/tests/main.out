*** -------------------------------------------------------------------------
*** Warning: UnitInterval has been deprecated in DOLFIN version 1.1.0.
*** The class UnitInterval has been replaced by UnitIntervalMesh.
*** -------------------------------------------------------------------------

*** -------------------------------------------------------------------------
*** Warning: UnitSquare has been deprecated in DOLFIN version 1.1.0.
*** The class UnitSquare has been replaced by UnitSquareMesh.
*** -------------------------------------------------------------------------

Solving linear system of size 3362 x 3362 (PETSc Krylov solver).
*** -------------------------------------------------------------------------
*** Warning: uint-valued MeshFunction has been deprecated in DOLFIN version 1.1.0.
*** Typename "uint" has been changed to "size_t".
*** -------------------------------------------------------------------------

Traceback (most recent call last):
  File "main.py", line 122, in <module>
    solver.assemble(nu,f,bc,q)
  File "/u/ifumagalli/tests/bin_mod/myschur.py", line 89, in assemble
    self.A, self.b = assemble_system(a, L, bc)
  File "/opt/libs/dolfin/1.2.0/gcc/4.8.0/lib/python2.7/site-packages/dolfin/fem/assembling.py", line 317, in assemble_system
    x0)
  File "/u/ifumagalli/tests/bin_mod/geometry.py", line 42, in eval
    tmp = self.q(x[0])
  File "/opt/libs/dolfin/1.2.0/gcc/4.8.0/lib/python2.7/site-packages/dolfin/functions/function.py", line 382, in __call__
    self.eval(values, x)
RuntimeError: 

*** -------------------------------------------------------------------------
*** DOLFIN encountered an error. If you are not able to resolve this issue
*** using the information listed below, you can ask for help at
***
***     https://answers.launchpad.net/dolfin
***
*** Remember to include the error message listed below and, if possible,
*** include a *minimal* running example to reproduce the error.
***
*** -------------------------------------------------------------------------
*** Error:   Unable to create intersection operator.
*** Reason:  IntersectionOperatorImplementation is not available, DOLFIN has been compiled without CGAL.
*** Where:   This error was encountered inside IntersectionOperator.cpp.
*** Process: 0
*** -------------------------------------------------------------------------

